### DISCUSSION TITLE
Diligent::DRAW_FLAG_DYNAMIC_RESOURCE_BUFFERS_INTACT

### DISCUSSION THREAD

**Discussion by RobertMtx** [DiligentGraphics/DiligentEngine#250] [Q&A] [answered]
My engine uses dynamic buffers pretty extensively. During 2D/particle/etc drawing, I essentially fill a dynamic buffer until the shader/material changes or the buffer runs out of space, then render them, remap, then repeat. But apart from dynamic drawing situations, almost all of my dynamic buffers are filled at once, then everything is rendered. Outside of the situation described above, almost all normal scene rendering happens by filling the buffers once, then rendering everything after. As a side note, do you think this non-2D process would be better served with non-dynamic buffers?

Anyway, based on the performance guide, it sounds like every call to `DrawIndexed()` may be performing some kind of expensive processing because of dynamic buffer use, even if they are not modified since the last call. So my question is - can I keep track of this situation with a single boolean variable?

1. When dynamic buffer memory is written to, mapped, or modified, I set this variable to true.
2. After every call to `DrawIndexed()` (or similar), I set the same variable to false.
3. When `DrawIndexed()` (or similar) is called, I send the `Diligent::DRAW_FLAG_DYNAMIC_RESOURCE_BUFFERS_INTACT` flag when the variable is false.

Would it be this simple? Or is there something I'm not understanding about how it works?

Edit: One other thing. In situations like this, is there a separate function that can be called instead of handling it this way? For example, is there a way I can send `Diligent::DRAW_FLAG_DYNAMIC_RESOURCE_BUFFERS_INTACT` to every draw call, but make sure I call some other specific function right before rendering modified dynamic buffer data?

**Comment by TheMostDiligent** [comment, marked as answer]
> As a side note, do you think this non-2D process would be better served with non-dynamic buffers?

Maybe. Hard to tell until you try.

> Anyway, based on the performance guide, it sounds like every call to DrawIndexed() may be performing some kind of expensive processing because of dynamic buffer use, even if they are not modified since the last call. 

This processing is not very expensive, but adds up.

> So my question is - can I keep track of this situation with a single boolean variable?

That sounds reasonable. Take a look at the asteroids sample. It is doing very similar thing.

> but make sure I call some other specific function right before rendering modified dynamic buffer data?

No, it is internal to the draw command and you should use the flag.

**Reply by RobertMtx** [reply]
Thanks, I appreciate the information. I didn't get much of a change when I started sending the flag, but I'm sure the improvement will grow as the number of draw commands does.

**Reply by TheMostDiligent** [reply]
Like I said, this does have some impact, but it is not high. You will see the difference if you run 10000+ draw calls.
